File: 052_ValidParentheses.cpp
/*
Problem: Valid Parentheses (Balanced Parentheses)
Description: Check if a string of parentheses (), {}, [] is valid and balanced.

Approach: Use stack. Push opening brackets, pop and match with closing brackets.
*/

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    bool isValid(string s) {
        stack<char> st;
        
        for (char c : s) {
            if (c == '(' || c == '{' || c == '[') {
                st.push(c);
            } else {
                if (st.empty()) return false;
                
                char top = st.top();
                if ((c == ')' && top == '(') ||
                    (c == '}' && top == '{') ||
                    (c == ']' && top == '[')) {
                    st.pop();
                } else {
                    return false;
                }
            }
        }
        
        return st.empty();
    }
};

int main() {
    Solution sol;
    string s = "()[]{}";
    
    cout << "Is Valid: " << (sol.isValid(s) ? "Yes" : "No") << endl;
    
    return 0;
}

/*
Time Complexity: O(n) where n is the length of string
Space Complexity: O(n) for the stack
*/

File: 052_ValidParentheses.py
"""
Problem: Valid Parentheses (Balanced Parentheses)
Description: Check if a string of parentheses (), {}, [] is valid and balanced.

Approach: Use stack. Push opening brackets, pop and match with closing brackets.
"""

class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        mapping = {')': '(', '}': '{', ']': '['}
        
        for char in s:
            if char in mapping:
                top_element = stack.pop() if stack else '#'
                
                if mapping[char] != top_element:
                    return False
            else:
                stack.append(char)
        
        return not stack

if __name__ == "__main__":
    sol = Solution()
    s = "()[]{}"
    
    print(f"Is Valid: {'Yes' if sol.isValid(s) else 'No'}")

"""
Time Complexity: O(n) where n is the length of string
Space Complexity: O(n) for the stack
"""
